tsfresh/feature_extraction/feature_calculators.py

```diff
@@ -427,27 +427,42 @@ def augmented_dickey_fuller(x, param):
 
     :param x: the time series to calculate the feature of
     :type x: numpy.ndarray
-    :param param: contains dictionaries {"attr": x} with x str, either "teststat", "pvalue" or "usedlag"
+    :param param: contains dictionaries {"attr": x, "autolag": y} with x str, either "teststat", "pvalue" or "usedlag"
+                  and with y str, either of "AIC", "BIC", "t-stats" or None.
     :type param: list
     :return: the value of this feature
     :return type: float
     """
-    res = None
-    try:
-        res = adfuller(x)
-    except LinAlgError:
-        res = np.NaN, np.NaN, np.NaN
-    except ValueError: # occurs if sample size is too small
-        res = np.NaN, np.NaN, np.NaN
-    except MissingDataError: # is thrown for e.g. inf or nan in the data
-        res = np.NaN, np.NaN, np.NaN
-
-    return [('attr_"{}"'.format(config["attr"]),
-                  res[0] if config["attr"] == "teststat"
-             else res[1] if config["attr"] == "pvalue"
-             else res[2] if config["attr"] == "usedlag" else np.NaN)
-            for config in param]
 
+    def compute_adf(autolag):
+        try:
+            return adfuller(x, autolag=autolag)
+        except LinAlgError:
+            return (np.NaN, np.NaN, np.NaN)
+        except ValueError: # occurs if sample size is too small
+            return (np.NaN, np.NaN, np.NaN)
+        except MissingDataError: # is thrown for e.g. inf or nan in the data
+            return (np.NaN, np.NaN, np.NaN)
+
+    cache = {}
+    res = []
+    for config in param:
+        autolag = config.get("autolag", "AIC")
+        if autolag not in cache:
+            cache[autolag] = compute_adf(autolag)
```

Comment:
You could move the last three lines also into the helper function and just call it here - would look a bit cleaner in my opinion

Reply:
> You could move the last three lines also into the helper function and just call it here - would look a bit cleaner in my opinion

OK. I fixed the helper function to cache the result automatically.